arbnode/espresso_delayed_message_fetcher.go

```diff
@@ -306,42 +229,123 @@ func (f *DelayedMessageFetcher) readDelayedMessage(seqNum uint64) (*DelayedInbox
-	if err != nil {
-		return fmt.Errorf("failed to put next hotshot block: %w", err)
+	// If we need to wait for confirmations,
+	// then get the latest block number - requiredBlockDepth
+	if d.waitForConfirmations {
+		latestBlockNumber, err := d.l1Reader.Client().BlockNumber(ctx)
+		if err != nil {
+			return 0, err
+		}
+		// Get the latest block - requiredBlockDepth
+		return latestBlockNumber - d.requiredBlockDepth, nil
 	}
 
-	return nil
+	// If no value is set, just use the latest block number
+	return d.l1Reader.Client().BlockNumber(ctx)
+}
+
+// getDelayedMessageCountAtBlock is a wrapper function for the delayedBridge.GetMessageCount function. This allows users of the DelayedMessageFetcher
+// to query for the message count at a block.
+func (f *DelayedMessageFetcher) getDelayedMessageCountAtBlock(blockNumber uint64) (uint64, error) {
+	count, err := f.delayedBridge.GetMessageCount(context.Background(), new(big.Int).SetUint64(blockNumber))
+	if err != nil {
+		return 0, err
+	}
+	return count, nil
 }
 
 /*
-Reads the current L1 block from the database.
+getDelayedMessagedInRange fetches all the delayed messages in the range [startBlock, endBlock]
+and stores them in the database
 */
-func readCurrentL1BlockFromDb(db ethdb.Database) (uint64, error) {
-	var blockNumber uint64
-	blockNumberBytes, err := db.Get([]byte(DelayedFetcherCurrentL1BlockKey))
-	if err != nil && !dbutil.IsErrNotFound(err) {
-		return 0, fmt.Errorf("failed to get next hotshot block: %w", err)
+func (d *DelayedMessageFetcher) getDelayedMessagesInRange(ctx context.Context, batch ethdb.Batch, startBlock uint64, endBlock uint64) error {
+
+	// Fetching the sequencer batches is important so that we can later parse the batch and get the sequencer batch data to store in the database
+	log.Debug("Looking for batches in range", "from", startBlock, "to", endBlock)
+	startBlockBigInt := big.NewInt(0).SetUint64(startBlock)
+	endBlockBigInt := big.NewInt(0).SetUint64(endBlock)
+
+	sequencerBatches, err := d.sequencerInbox.LookupBatchesInRange(ctx, startBlockBigInt, endBlockBigInt)
+	if err != nil {
+		return err
 	}
-	if blockNumberBytes != nil {
-		err = rlp.DecodeBytes(blockNumberBytes, &blockNumber)
+	log.Debug("Sequencer batches found", "sequencerBatches", sequencerBatches)
+
+	log.Debug("Looking for delayed messages from range", "from", startBlock, "to", endBlock)
+	msgs, err := d.delayedBridge.LookupMessagesInRange(ctx, big.NewInt(0).SetUint64(startBlock), big.NewInt(0).SetUint64(endBlock), func(batchNum uint64) ([]byte, error) {
+		if len(sequencerBatches) > 0 && batchNum >= sequencerBatches[0].SequenceNumber {
+			idx := batchNum - sequencerBatches[0].SequenceNumber
+			if idx < uint64(len(sequencerBatches)) {
+				return sequencerBatches[idx].Serialize(ctx, d.l1Reader.Client())
+			}
+			return nil, fmt.Errorf("failed to get sequencer batch data: %w", err)
+		} else {
+			return nil, fmt.Errorf("failed to get sequencer batch data: %w", err)
+		}
```

Comment:
```suggestion
		}
		return nil, fmt.Errorf("failed to get sequencer batch data: %w", err)
```

`else` statement is redundant.